### ISSUE TITLE
Use of deprecated core methods around JRuby ScriptingContainer and Timestamp

### ISSUE THREAD

**Issue by dmikurube** [embulk/embulk#745]
Some constructors of `TimestampFormatter` and `TimestampParser` have required JRuby `ScriptingContainer` while the JRuby instance has not been used actually. Throughout the process of un-JRuby-ing the bootstrap, these constructors are deprecated in #744, and are to be removed in the future (maybe with the major update v0.9).

A deprecation warning message is printed if a plugin uses the deprecated constructors in user's run since v0.8.29. Comments are welcome in this issue to report plugins using the constructors.

There are several ways for plugins and plugin developers to catch up.

----

#### `TimestampFormatter`
* `TimestampFormatter(String, FormatterTask task)` (deprecated since v0.6.14)
* `TimestampFormatter(ScriptingContainer jruby, String format, DateTimeZone timeZone)` (deprecated since v0.8.29)

#### `TimestampParser`
* `TimestampParser(String format, ParserTask task)` (deprecated since v0.6.14)
* `TimestampParser(ScriptingContainer jruby, String format, DateTimeZone defaultTimeZone)` (deprecated since v0.8.29)
* `TimestampParser(ScriptingContainer jruby, String format, DateTimeZone defaultTimeZone, String defaultDate)` (deprecated since v0.8.29)

### Advice for plugin developers:

It is acceptable to ask users to ignore the deprecation messages for a while, but it is recommended to catch up with the deprecation within 3 months (~ November, 2017). These constructors would be finally removed, in v0.9 in the shortest, or in around 6 months.

#### `TimestampFormatter`

The safest way is to use `TimestampFormatter(Task task, Optional<? extends TimestampColumnOption> columnOption)` instead because this constructor has been long-lived since v0.6.14. It can be implemented without a migration period although it may need some efforts as the method signature is so different.

Another option is to use `TimestampFormatter(String format, DateTimeZone timeZone)`. It is easy to implement, but it runs only with v0.8.29+. You may need to wait for your users to update their Embulk core to 0.8.29+ (with ignoring the deprecation message), and then update your plugin to call this constructor.

#### `TimestampParser`

Similar to `TimestampFormatter`. The safest way is to use `TimestampParser(Task task, TimestampColumnOption columnOption)`.

Other options only for 0.8.29+ are to use `TimestampParser(String format, DateTimeZone defaultTimeZone)` or `TimestampParser(String format, DateTimeZone defaultTimeZone, String defaultDate)`.

**Comment by sonots** [comment]
Are there any easy ways to create instances of TimestampColumnOption?

With a lucky case https://github.com/sonots/embulk-filter-column/pull/22, I could use `TimestampParser(Task task, TimestampColumnOption columnOption)` as:

```java
public class ColumnFilterPlugin implements FilterPlugin
{
    interface ColumnOption extends Task, TimestampParser.TimestampColumnOption
    {
        // TimestampParser.TimestampColumnOption defines timezone, format, and date
    }

    interface PluginTask extends Task, TimestampParser.Task
    {
        // TimestampParser.Task defines default_timestamp_format, default_timezone, and default_date

        @Config("columns")
        @ConfigDefault("[]")
        public List<ColumnOption> getColumns();
    }
```

```java
for (ColumnOption clumnOption : task.getColumns()) {
    TimestampParser parser = new TimestampParser(task, columnOption);
}
```

However, with unlucky case https://github.com/sonots/embulk-filter-typecast, I wanted to write codes as:

```java
public class TypecastFilterPlugin implements FilterPlugin
{
    public interface ColumnConfig extends Task,
        TimestampParser.TimestampColumnOption, TimestampFormatter.TimestampColumnOption
    {
    }
```

because I use `format` and `timezone` parameter for both formatter and parser in this plugin, and resulted in 

```
Error: com.fasterxml.jackson.databind.JsonMappingException: Multiple entries with same key: JRuby=public abstract org.jruby.embed.ScriptingContainer org.embulk.spi.time.TimestampFormatter$Task.getJRuby()
and JRuby=public abstract org.jruby.embed.ScriptingContainer org.embulk.spi.time.TimestampParser$Task.getJRuby()
```

**Comment by sonots** [comment]
If I can manually create instances of `TimestampFormatterr.TimestampColumnOption`, I should be able to avoid the above error with codes such as:

```java
public class TypecastFilterPlugin implements FilterPlugin
{
    public interface ColumnOption extends Task, TimestampParser.TimestampColumnOption
    {
    }
```

```java
formatterColumnnOptions = createTimestampFormatterTimestampColumnOptions(task.getColumns()); // how?
for (ColumnOption clumnOption : formatterColumnnOptions) {
    TimestampFormatter formatter = new TimestampFormatter(task, Optional.of(columnOption));
}
```

**Comment by sonots** [comment]
With another unlucky case https://github.com/sonots/embulk-formatter-single_value, I have different task option names `timestamp_format`, and `timezone` rather than `default_timestamp_format`, and `default_timezone` and I do not have any ColumnOptions.

How can I use `TimestampFormatter(Task task, Optional<? extends TimestampColumnOption> columnOption)` in this case? Are there any easy ways to create instances of TimestampColumnOption?

**Comment by dmikurube** [comment]
@sonots Thanks for taking care of this issue on your plugins!

`TimestampColumnOption` itself is a simple (annotated) interface. So, the following simple example works actually (though it's like a bit hacky) :

```java
private class TimestampColumnOptionImpl implements TimestampParser.TimestampColumnOption {
    public TimestampColumnOptionImpl(DateTimeZone dateTimeZone, String format, String date) {
        this.dateTimeZone = dateTimeZone;
        this.format = format;
        this.date = date;
    }
    @Override
    public Optional<DateTimeZone> getTimeZone() {
        return Optional.fromNullable(this.dateTimeZone);
    }
    @Override
    public Optional<String> getFormat() {
        return Optional.fromNullable(this.format);
    }
    @Override
    public Optional<String> getDate() {
        return Optional.fromNullable(this.date);
    }
    private final DateTimeZone dateTimeZone;
    private final String format;
    private final String date;
}
```

Or, more properly like Embulk's config (maybe more recommended):

```java
private interface TimestampColumnOptionIntl extends org.embulk.config.Task, TimestampParser.TimestampColumnOption {}
```

```java
// ...
ConfigSource configSource = Exec.newConfigSource();
configSource.set("timeZone", Optional.fromNullable(timezone));
// ...
TimestampColumnOptionIntl timestampColumnOption = configSource.loadConfig(TimestampColumnOptionIntl.class);
```

If you use this, it may be nice to have a TODO comment like: `Replace this to the other constructor after an enough interval after Embulk 0.8.29.`

**Comment by sonots** [comment]
Thank you! I will try it later.

**Comment by sonots** [comment]
I wrote a document about this http://qiita.com/sonots/items/ac56aa58f068308e120f (Japanese).

**Comment by dmikurube** [comment]
@sonots Thanks for your contribution!

Just in case, the reason why `ConfigSource` is more recommended is that it's more robust. If we have changes in `TimestampParser.Task` in the future (though we don't want to have), the `implements TimestampParser.Task` may not work then. If that is okay, `implements TimestampParser.Task` is fine.

**Comment by sonots** [comment]
I got you. But, I rather felt uneasy for `ConfigSource` because of the looseness.
When parameter names of `TimestampParser.Task` is changed, compilation still passes, but it actually does not work. We can not easily notice it.

Because of this reason also, I recommended `implements TimestampParser.Task` way and used it in my plugins.

----

For example, the compilation passes for below codes, but this is actually wrong. Developers can not easily notice it.

```
ConfigSource configSource = Exec.newConfigSource();
configSource.set("timeZone", Optional.fromNullable(timezone));
```
